### PR TITLE
Icu 11022 update dynamic credentials in sb form

### PR DESCRIPTION
**Pull request by DhariniJeeva**
:tickets: [Jira ticket](https://hashicorp.atlassian.net/browse/ICU-11022)

## Description

This PR allows credentials to switch between static and dynamic.

## Screenshots (if appropriate):

Screenshot 2023-09-15 at 11 20 37 AM

## How to Test

:technologist: [Admin preview](PATH_TO_FEATURE)

:desktop_computer: [Desktop preview](PATH_TO_FEATURE)

## Checklist:

- [x] I have added before and after screenshots for UI changes
- ~[ ] I have added JSON response output for API changes~
- ~[ ] I have added steps to reproduce and test for bug fixes in the description~
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by dvonanderson — APPROVED** [review]
LGTM

**Review by cameronperera — APPROVED** [review]

**Review by calcaide — APPROVED** [review]
LGTM!! Nice work!! killing the naming game with const like `fieldsByCredentialType` there is no more need for context, awesome!!

File: addons/api/addon/generated/models/storage-bucket.js [hunk 1]
```diff
@@ -73,12 +73,6 @@ export default class GeneratedStorageBucketModel extends BaseModel {
   })
   region;
 
-  @attr('string', {
-    readOnly: true,
-    isNestedAttribute: true,
-  })
-  secrets_hmac;
```
Comment (ZedLi) [inline addons/api/addon/generated/models/storage-bucket.js:80]: Do we not get a `secrets_hmac` back anymore? Why was this deleted?

File: addons/api/addon/serializers/storage-bucket.js [hunk 2]
```diff
@@ -4,29 +4,62 @@
...
+    }
     return serialized;
   }
 
+  serializeAttribute(snapshot, json, key, attribute) {
+    const value = super.serializeAttribute(...arguments);
+    const { credentialType } = snapshot.record;
+    const { options } = attribute;
+    // This deletes any fields that don't belong to the record's credential type
+    if (options.isNestedAttribute && json.attributes) {
```
Comment (ZedLi) [inline addons/api/addon/serializers/storage-bucket.js:45]: Does this work correctly for `access_key_id` and `secret_access_key`? They're only annotated as `isNestedSecret` so wouldn't we skip over them?

File: addons/api/addon/serializers/storage-bucket.js [hunk 3]
```diff
@@ -4,29 +4,62 @@
...
-    const normalized = super.normalize(typeClass, normalizedHash, ...rest);
 
     // Remove secret fields so we don't track them after creating/updating
-    normalized.data.attributes.access_key_id = null;
-    normalized.data.attributes.secret_access_key = null;
-    return normalized;
+    if (normalizedHash?.secrets?.access_key_id) {
+      delete normalizedHash.secrets.access_key_id;
+      delete normalizedHash.secrets.secret_access_key;
+    }
```
Comment (ZedLi) [inline addons/api/addon/serializers/storage-bucket.js:62]: Why did this change? This is for normalization and I believe we don't store anything as a `secrets` object until serialization so we never actually represent it as `secrets` but just attributes.

Also for normalization we have to set it to `null` as ember will interpret a missing attribute (which is what `delete` will do) as no change so the secrets never get deleted.

File: addons/api/tests/unit/serializers/storage-bucket-test.js [hunk 4]
```diff
@@ -161,6 +154,57 @@ module('Unit | Serializer | storage bucket', function (hooks) {
     assert.deepEqual(serializedRecord, expectedResult);
   });
 
+  test('it serializes when updating a storage bucket with dynamic credential correctly', async function (assert) {
```
Comment (ZedLi) [inline addons/api/tests/unit/serializers/storage-bucket-test.js:157]: Can we add values for the `access_key_id` and `secret_access_key` to confirm they properly get set to null?

File: addons/api/tests/unit/serializers/storage-bucket-test.js [hunk 5]
```diff
@@ -125,13 +123,8 @@ module('Unit | Serializer | storage bucket', function (hooks) {
           region: 'eu-west-1',
           access_key_id: 'foobars',
           secret_access_key: 'testing',
-          secrets_hmac: 'hmac',
           disable_credential_rotation: true,
           version: 1,
-          role_arn: null,
-          role_external_id: null,
-          role_session_name: null,
-          role_tags: [],
```
Comment (ZedLi) [inline addons/api/tests/unit/serializers/storage-bucket-test.js:134]: Can we also set values for these to confirm the serialization correctly sets them to null?

File: addons/api/addon/generated/models/storage-bucket.js (hunk 1 above)
Comment (DhariniJeeva) [reply addons/api/addon/generated/models/storage-bucket.js:80]: Oops no this is still needed
